actors/builtin/miner/expiration_queue.go

```diff
@@ -0,0 +1,636 @@
+	rescheduledPower := NewPowerPairZero()
+
+	var es ExpirationSet
+	if err := q.Array.ForEach(&es, func(e int64) error {
+		epoch := abi.ChainEpoch(e)
+		if epoch <= q.quant.QuantizeUp(faultExpiration) {
+			// Regardless of whether the sectors were expiring on-time or early, all the power is now faulty.
+			// Pledge is still on-time.
+			es.FaultyPower = es.FaultyPower.Add(es.ActivePower)
+			es.ActivePower = NewPowerPairZero()
+			if err := q.mustUpdate(epoch, &es); err != nil {
+				return err
+			}
+		} else {
+			rescheduedEpochs = append(rescheduedEpochs, uint64(epoch))
+			rescheduledSectors = append(rescheduledSectors, es.OnTimeSectors, es.EarlySectors)
+			rescheduledPower = rescheduledPower.Add(es.ActivePower)
+			rescheduledPower = rescheduledPower.Add(es.FaultyPower)
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// Add rescheduled sectors to new expiration as early-terminating and faulty.
+	allRescheduled, err := bitfield.MultiMerge(rescheduledSectors...)
+	if err != nil {
+		return xerrors.Errorf("failed to merge rescheduled sectors: %w", err)
+	}
+	noOnTimeSectors := abi.NewBitField()
+	noActivePower := NewPowerPairZero()
+	noOnTimePledge := abi.NewTokenAmount(0)
+	if err = q.add(faultExpiration, noOnTimeSectors, allRescheduled, noActivePower, rescheduledPower, noOnTimePledge); err != nil {
+		return err
+	}
+
+	// Trim the rescheduled epochs from the queue.
+	if err = q.BatchDelete(rescheduedEpochs); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Removes sectors from any queue entries in which they appear that are earlier then their scheduled expiration epoch,
+// and schedules them at their expected termination epoch.
+// Pledge for the sectors is re-added as on-time.
+// Power for the sectors is changes from faulty to active (whether rescheduled or not).
+// Returns the newly-recovered power. Fails if any sectors are not found in the queue.
+func (q ExpirationQueue) RescheduleRecovered(sectors []*SectorOnChainInfo, ssize abi.SectorSize) (PowerPair, error) {
+	remaining := make(map[abi.SectorNumber]struct{}, len(sectors))
+	for _, s := range sectors {
+		remaining[s.SectorNumber] = struct{}{}
+	}
+
+	// Traverse the expiration queue once to find each recovering sector and remove it from early/faulty there.
+	var sectorsRescheduled []*SectorOnChainInfo
+	recoveredPower := NewPowerPairZero()
+	if err := q.traverseMutate(func(epoch abi.ChainEpoch, es *ExpirationSet) (changed, keepGoing bool, err error) {
+		onTimeSectors, err := es.OnTimeSectors.AllMap(SectorsMax)
+		if err != nil {
+			return false, false, err
+		}
+		earlySectors, err := es.EarlySectors.AllMap(SectorsMax)
+		if err != nil {
+			return false, false, err
+		}
+
+		// This loop could alternatively be done by constructing bitfields and intersecting them, but it's not
+		// clear that would be much faster (O(max(N, M)) vs O(N+M)).
+		// If faults are correlated, the first queue entry likely has them all anyway.
```

Comment:
The first queue entry _with faults_ likely has all of them. We could still process 13 epochs with other expirations first. But I agree that optimizing this is probably not worth it.